pkg/client/ssa_client_test.go

```diff
@@ -0,0 +1,200 @@
+package client_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/codeready-toolchain/toolchain-common/pkg/client"
+	"github.com/codeready-toolchain/toolchain-common/pkg/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestSsaClient(t *testing.T) {
+	t.Run("ApplyObject", func(t *testing.T) {
+		t.Run("creates", func(t *testing.T) {
+			// given
+			cl, acl := NewTestSsaApplyClient(t)
+			obj := &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "obj",
+					Namespace: "default",
+				},
+			}
+
+			// when
+			require.NoError(t, acl.ApplyObject(context.TODO(), obj))
+
+			// then
+			inCluster := &corev1.ConfigMap{}
+			require.NoError(t, cl.Get(context.TODO(), runtimeclient.ObjectKeyFromObject(obj), inCluster))
+		})
+		t.Run("updates", func(t *testing.T) {
+			// given
+			obj := &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "obj",
+					Namespace: "default",
+				},
+				Data: map[string]string{"a": "b"},
+			}
+			cl, acl := NewTestSsaApplyClient(t, obj)
+
+			updated := obj.DeepCopy()
+			updated.Data["a"] = "c"
+
+			// when
+			require.NoError(t, acl.ApplyObject(context.TODO(), updated))
+			inCluster := &corev1.ConfigMap{}
+			require.NoError(t, cl.Get(context.TODO(), runtimeclient.ObjectKeyFromObject(obj), inCluster))
+
+			// then
+			assert.Equal(t, "c", inCluster.Data["a"])
```

Comment:
not sure I'm following how we are testing updates - I mean the obj is not actually being first created ( with the fake client)  and than updated. You're just updating the "local" version  of the object. But I'm sure I'm missing something.

Reply:
ok, I see. I didn't saw that passing the obj to the constructor actually adds it to the fake client. 
I was expecting something like:
- I apply one object with some data in one test
- In a subtest I modify the object and apply it again and than I check that the object has all the changes